### PR TITLE
Fix recursive redirects.

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/137)

### CODE REVIEW

**Comment by SomeoneToIgnore** [comment]
*[src/router.js, line 25 at r1](https://reviewable.io/reviews/vaadin/vaadin-router/137#-LEBWR8wECCzo6Gtt4Og:-LEBWR8wECCzo6Gtt4Oh:b-geyq5s) ([raw file](https://github.com/vaadin/vaadin-router/blob/2eb8d642e81ca15f6f13ee9a38cb36bb79a68c92/src/router.js#L25)):*
> ```JavaScript
> }
> 
> function runCallbackIfPossible(callback, context) {
> ```

Creating this function may seem to be an overkill now, but it's here for https://github.com/vaadin/vaadin-router/pull/138 and for the case when we need to run `onBeforeEnter` and `onBeforeLeave` methods.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/137)*

**Comment by SomeoneToIgnore** [comment]
*[test/router/dynamic.redirect.spec.html, line 286 at r1](https://reviewable.io/reviews/vaadin/vaadin-router/137#-LEBWhTeDplyYBosr_jw:-LEBWhTeDplyYBosr_jx:ba4y3j5) ([raw file](https://github.com/vaadin/vaadin-router/blob/2eb8d642e81ca15f6f13ee9a38cb36bb79a68c92/test/router/dynamic.redirect.spec.html#L286)):*
> ```HTML
>       });
> 
>       describe('recursive redirects', () => {
> ```

2 more methods (for `onBeforeEnter` and `onBeforeLeave`) will be added in future.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/137)*

**Comment by web-padawan** [comment]
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[src/router.js, line 26 at r1](https://reviewable.io/reviews/vaadin/vaadin-router/137#-LEF80kv8y9vU22JES9L:-LEF80kv8y9vU22JES9M:b-giqwdv) ([raw file](https://github.com/vaadin/vaadin-router/blob/24ceb6ae6ff7becd3dc3fcb16d762912e1fa6dde/src/router.js#L26)):*
> ```JavaScript
> 
> function runCallbackIfPossible(callback, context) {
>   if (callback && typeof callback === 'function') {
> ```

the typeof condition check if sufficient, no falsey values can bypass it, unlike the "object" which will satisfy `null` value.

---

*[src/router.js, line 34 at r1](https://reviewable.io/reviews/vaadin/vaadin-router/137#-LEF89Am7UqTCcHmUcpC:-LEF7-TmBtzc7iSQDGby:bfd4n1y) ([raw file](https://github.com/vaadin/vaadin-router/blob/24ceb6ae6ff7becd3dc3fcb16d762912e1fa6dde/src/router.js#L34)):*
> ```JavaScript
> }
> 
> function redirectsToCurrentStateSecondTimeInARow(context, redirectData) {
> ```

BTW can we shorten this to something like `isCyclicRedirect` or similar?
I would prefer using "is" for methods returning Boolean value.

This is just my personal opinion and not a blocker.

---

*[src/router.js, line 192 at r1](https://reviewable.io/reviews/vaadin/vaadin-router/137#-LEF7gJi3vkCGm7snDay:-LEF7gJj9gqAQ1BzlzGF:b-58b3r5) ([raw file](https://github.com/vaadin/vaadin-router/blob/24ceb6ae6ff7becd3dc3fcb16d762912e1fa6dde/src/router.js#L192)):*
> ```JavaScript
>       if (!redirectsToCurrentStateSecondTimeInARow(context, redirectTarget.redirect)) {
>         return redirectTarget;
>       }
> ```

Consider adding the "else" block with the short warning.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/137)*

**Comment by vlukashov** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[test/router/dynamic.redirect.spec.html, line 298 at r1](https://reviewable.io/reviews/vaadin/vaadin-router/137#-LEF8UKHE4DjiSavqxXK:-LEF8UKI5JaBPyW4cI45:b-vlq32a) ([raw file](https://github.com/vaadin/vaadin-router/blob/24ceb6ae6ff7becd3dc3fcb16d762912e1fa6dde/test/router/dynamic.redirect.spec.html#L298)):*
> ```HTML
>         it('recursive redirect defined with property works and activates the path once', async() => {
>           router.setRoutes([
>             {path: '/', redirect: '/', component: 'x-home-view'}
> ```

How does that work together with the warning `Route with path '/' has both 'redirect' and 'component' properties specified. Latter will never evaluated since 'redirect' will be executed earlier.`? In effect, the `component` property does get evaluated.

How about throwing an error when a cyclic redirect is detected?

---

*[test/router/dynamic.redirect.spec.html, line 306 at r1](https://reviewable.io/reviews/vaadin/vaadin-router/137#-LEF8n1m9DhL0CWkmvfM:-LEF8n1m9DhL0CWkmvfN:b-1zhc0z) ([raw file](https://github.com/vaadin/vaadin-router/blob/24ceb6ae6ff7becd3dc3fcb16d762912e1fa6dde/test/router/dynamic.redirect.spec.html#L306)):*
> ```HTML
>         });
> 
>         it('recursive redirect caused by action works and activates the path once', async() => {
> ```

Please add a test case for a cyclic redirect `/a` -> `/b` -> `/a`. I'd expect this to throw at the run time.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/137)*

**Comment by SomeoneToIgnore** [comment]
Sorry, I have to 

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://reviewable.io/reviews/vaadin/vaadin-router/137#-LEGf7tN0o2xeBMRbQTM:-LEGf7tN0o2xeBMRbQTN:b3mnapu) (no related file):*
I've just realised a few things:

1. This PR is not needed for my functionality anymore, since after the refactoring I've made it works fine without this feature.

1. This feature does not cover `/a -> /b -> /a` case but this case does not work for us either.

1. We cannot catch `/a -> /b -> /c -> /d -> /a` case without either introducing some stack that remembers our redirects or some other technique.
If we solve this problem, `/a -> /a` case will be solved automatically since it's just less general case.

Hence I'll close the PR, sorry for taking your time.

Tickets created: 
* https://github.com/vaadin/vaadin-router/issues/144
* https://github.com/vaadin/vaadin-router/issues/145

---

*[test/router/dynamic.redirect.spec.html, line 306 at r1](https://reviewable.io/reviews/vaadin/vaadin-router/137#-LEF8n1m9DhL0CWkmvfM:-LEGet4R6S7_FJfT1Q1N:b-8svu5m) ([raw file](https://github.com/vaadin/vaadin-router/blob/24ceb6ae6ff7becd3dc3fcb16d762912e1fa6dde/test/router/dynamic.redirect.spec.html#L306)):*
<details><summary><i>Previously, vlukashov (Viktor Lukashov) wrote…</i></summary><blockquote>

Please add a test case for a cyclic redirect `/a` -> `/b` -> `/a`. I'd expect this to throw at the run time.
</blockquote></details>

I've totally forgotten about this use case and after I've tried the positive scenario it did not work for me: 

```
router.setRoutes([
            {path: '/a', redirect: '/b'},
            {path: '/b', redirect: '/c'},
            // {path: '/c', redirect: '/a', component: 'x-home-view'},
            {path: '/c', component: 'x-home-view'},
          ]);

          await router.render('/a');
```

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/137#-:-LEGenFqB9nk2c3F_kzw:bjjjxu8)*

**Comment by SomeoneToIgnore** [comment]
-
